beacon-chain/sync/initial-sync/blocks_fetcher_utils.go

```diff
@@ -119,6 +141,139 @@ func (f *blocksFetcher) nonSkippedSlotAfter(ctx context.Context, slot uint64) (u
+// ones that feature blocks from alternative branches. Once found, peer is further queried
+// to find common ancestor slot. On success, all obtained blocks and peer is returned.
+func (f *blocksFetcher) findFork(ctx context.Context, slot uint64) (*forkData, error) {
+	ctx, span := trace.StartSpan(ctx, "initialsync.findFork")
+	defer span.End()
+
+	// Safe-guard, since previous epoch is used when calculating.
+	slotsPerEpoch := params.BeaconConfig().SlotsPerEpoch
+	if slot < slotsPerEpoch*2 {
+		return nil, fmt.Errorf("slot is to low to backtrack, min. expected %d", slotsPerEpoch*2)
+	}
+
+	// The current slot's epoch must be after the finalization epoch,
+	// triggering backtracking on earlier epochs is unnecessary.
+	finalizedEpoch := f.chain.FinalizedCheckpt().Epoch
+	epoch := helpers.SlotToEpoch(slot)
+	if epoch <= finalizedEpoch {
+		return nil, errors.New("slot is not after the finalized epoch, no backtracking is necessary")
+	}
+	// Update slot to the beginning of the current epoch (preserve original slot for comparison).
+	slot, err := helpers.StartSlot(epoch)
+	if err != nil {
+		return nil, err
+	}
+
+	// Select peers that have higher head slot, and potentially blocks from more favourable fork.
+	// Exit early if no peers are ready.
+	_, peers := f.p2p.Peers().BestNonFinalized(1, epoch+1)
+	if len(peers) == 0 {
+		return nil, errNoPeersAvailable
+	}
+	f.rand.Shuffle(len(peers), func(i, j int) {
+		peers[i], peers[j] = peers[j], peers[i]
+	})
+
+	// Query all found peers, stop on peer with alternative blocks, and try backtracking.
+	for i, pid := range peers {
+		log.WithFields(logrus.Fields{
+			"peer": pid,
+			"step": fmt.Sprintf("%d/%d", i+1, len(peers)),
+		}).Debug("Searching for alternative blocks")
+		fork, err := f.findForkWithPeer(ctx, pid, slot)
+		if err != nil {
+			log.WithFields(logrus.Fields{
+				"peer":  pid,
+				"error": err.Error(),
+			}).Debug("No alternative blocks found for peer")
+			continue
+		}
+		return fork, nil
+	}
+	return nil, errNoPeersWithAltBlocks
+}
+
+// findForkWithPeer loads some blocks from a peer in an attempt to find alternative blocks.
+func (f *blocksFetcher) findForkWithPeer(ctx context.Context, pid peer.ID, slot uint64) (*forkData, error) {
+	// Locate non-skipped slot, supported by a given peer (can survive long periods of empty slots).
+	// When searching for non-empty slot, start an epoch earlier - for those blocks we
+	// definitely have roots. So, spotting a fork will be easier. It is not a problem if unknown
+	// block of the current fork is found: we are searching for forks when FSMs are stuck, so
+	// being able to progress on any fork is good.
+	pidState, err := f.p2p.Peers().ChainState(pid)
+	if err != nil {
+		return nil, fmt.Errorf("cannot obtain peer's status: %w", err)
+	}
+	slotsPerEpoch := params.BeaconConfig().SlotsPerEpoch
+	nonSkippedSlot, err := f.nonSkippedSlotAfterWithPeersTarget(
+		ctx, slot-slotsPerEpoch, []peer.ID{pid}, helpers.SlotToEpoch(pidState.HeadSlot))
```

Comment:
`slot-slotsPerEpoch` , this would be an overflow in the event `slot` is less than 32

Reply:
I will add an extra check to this function (it is used from inside `findFork` where I do check that slot is higher that `slotsPerEpoch * 2`, but you never know from where a function is going to be called - so, better safe than sorry).